platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonFactory.java

```diff
@@ -29,17 +29,20 @@
 public class WorkerDaemonFactory implements WorkerFactory {
     private final WorkerDaemonClientsManager clientsManager;
     private final BuildOperationRunner buildOperationRunner;
+    private final WorkerDaemonClientCancellationHandler workerDaemonClientCancellationHandler;
 
-    public WorkerDaemonFactory(WorkerDaemonClientsManager clientsManager, BuildOperationRunner buildOperationRunner) {
+    public WorkerDaemonFactory(WorkerDaemonClientsManager clientsManager, BuildOperationRunner buildOperationRunner, WorkerDaemonClientCancellationHandler workerDaemonClientCancellationHandler) {
         this.clientsManager = clientsManager;
         this.buildOperationRunner = buildOperationRunner;
+        this.workerDaemonClientCancellationHandler = workerDaemonClientCancellationHandler;
     }
 
     @Override
     public BuildOperationAwareWorker getWorker(WorkerRequirement workerRequirement) {
         return new AbstractWorker(buildOperationRunner) {
             @Override
             public DefaultWorkResult execute(IsolatedParametersActionExecutionSpec<?> spec, BuildOperationRef parentBuildOperation) {
+                workerDaemonClientCancellationHandler.start();
```

Comment:
Could you explain the importance of this? IIUC, this is the magic that makes it so we only kill workers when a build is canceled if that build uses any worker.

Reply:
The `WorkerDaemonClientManager` is daemon-scoped (GUH-scoped to be precise).  `WorkerDaemonFactory` is project-scoped.  The build cancellation token is session-scoped, however.  We don't have any meaningful worker-related services that are session-scoped, so we need something in the lifecycle to step in at some point and register a the `WorkerDaemonClientCancellationHandler` with the right cancellation token for the current build session.  There is a session listener callback (which we use in `WorkerDaemonClientsManager`) but there's no good way to get the cancellation token when that fires.  On the other hand, the current build session doesn't need to kill workers on cancellation if it doesn't use any workers (i.e. if it doesn't use a worker, then it can't leave it in an unsafe state when cancelled).  The code you referenced is called whenever something (anything) tries to get (or start) a worker daemon.  So, at that point, we know a worker is in use - this then notifies the handler that it should register a callback with the cancellation token for the session.  If the handler has already been registered (i.e. `start()` has already been called during this session) it's just ignored.  We could name the method on `WorkerDaemonClientCancellationHandler` differently, I suppose (`notifyClientInUse()` or some such).  Does that answer your question?

Reply:
Ah, sorry.  That's a good explanation. I meant, could you leave a comment about this in the code :)

Reply:
The other options here would be to 1) always stop worker daemons on cancellation, regardless of if any were used or not or 2) create some mechanism to track when a worker is actually executing a work item and only stop worker daemons if a cancellation occurs while work execution is in progress.  These are the two extremes of optimization, I think.  I opted for something simple somewhere in the middle.